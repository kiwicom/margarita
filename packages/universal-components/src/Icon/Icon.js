// @flow

import * as React from 'react';
import { Text as RNText } from 'react-native';
import { defaultTokens } from '@kiwicom/orbit-design-tokens';

import { Text } from '../Text';
import {
  StyleSheet,
  type PlatformStyleObjectType,
} from '../PlatformStyleSheet';
import iconsMap from './icons.json';
import type { Props } from './IconTypes';

const sanitizeIconCharacter = (iconCharacter: string) => {
  if (/^E(.{3})$/.test(iconCharacter)) {
    return String.fromCharCode(parseInt(iconCharacter, 16));
  }
  return iconCharacter;
};

export default function Icon({
  name,
  color = '#46515e',
  size = 'medium',
  style,
}: Props) {
  if (name == null) {
    return <Text style={[styles[size], { color }, style]}>?</Text>;
  }

  const icon = iconsMap[name];
  if (!icon) {
    return <Text style={[styles[size], { color }, style]}>?</Text>;
  }

  return (
    <RNText style={[styles.icon, styles[size], { color }, style]}>
      {sanitizeIconCharacter(icon.character)}
    </RNText>
  );
}

const getSizeStyle = (size: number): $Values<PlatformStyleObjectType> => ({
  fontSize: size,
  width: size,
  height: size,
  lineHeight: size,
});

const styles = StyleSheet.create({
  icon: {
    fontFamily: 'orbit-icons',
    android: {
      includeFontPadding: false,
      textAlignVertical: 'center',
    },
  },
  /* eslint-disable react-native/no-unused-styles */
  small: getSizeStyle(parseFloat(defaultTokens.widthIconSmall)),
  medium: getSizeStyle(parseFloat(defaultTokens.widthIconMedium)),
  large: getSizeStyle(parseFloat(defaultTokens.widthIconLarge)),
  /* eslint-enable */
});
