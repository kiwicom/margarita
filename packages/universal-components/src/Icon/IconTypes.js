// @flow

import type { StylePropType } from '../PlatformStyleSheet/StyleTypes';
import type { IconNameType } from '../types/_generated-types';

type Size = 'small' | 'medium' | 'large';

export type Props = {|
  +name?: IconNameType,
  size?: Size,
  color?: string,
  style?: StylePropType,
|};
