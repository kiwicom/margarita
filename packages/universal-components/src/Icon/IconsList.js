// @flow strict

import * as React from 'react';
import { FlatList, View } from 'react-native';

import { Text } from '../Text';
import { StyleSheet } from '../PlatformStyleSheet';
import iconsMap from './icons.json';

import { Icon } from '.';

const keyExtractor = item => item;
const renderItem = ({ item }) => (
  <View key={item} style={styles.item}>
    <Text>{item}</Text>
    <Icon name={item} size="large" />
  </View>
);

export default function IconsList() {
  return (
    <FlatList
      data={Object.keys(iconsMap)}
      keyExtractor={keyExtractor}
      showsVerticalScrollIndicator={false}
      contentContainerStyle={styles.list}
      numColumns={2}
      renderItem={renderItem}
    />
  );
}

const styles = StyleSheet.create({
  list: {
    paddingVertical: 10,
    paddingHorizontal: 10,
  },
  item: {
    flex: 1,
    alignItems: 'center',
    paddingHorizontal: 20,
    paddingVertical: 10,
  },
});
