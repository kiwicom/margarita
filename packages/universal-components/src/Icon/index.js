// @flow

export { default as Icon } from './Icon';
