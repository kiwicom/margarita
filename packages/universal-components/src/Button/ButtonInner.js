// @flow

import * as React from 'react';
import { View } from 'react-native';
import { defaultTokens } from '@kiwicom/orbit-design-tokens';

import { Icon } from '../Icon';
import { StyleSheet } from '../PlatformStyleSheet';
import ButtonTitle from './ButtonTitle';
import type { ButtonType } from './ButtonTypes';
import { textColor, wrapperColor } from './styles';

type Props = {|
  +children?: React.Node,
  +disabled?: boolean,
  +type?: ButtonType,
  +leftIcon?: React.Element<typeof Icon> | null,
  +rightIcon?: React.Element<typeof Icon> | null,
  +testID?: string,
  +sublabel?: React.Node,
  +label?: React.Node,
  +circled?: boolean,
|};

export default function ButtonInner({
  disabled = false,
  type = 'primary',
  testID: testIDProps,
  children,
  leftIcon: originalLeftIcon,
  rightIcon: originalRightIcon,
  sublabel,
  label,
  circled,
}: Props) {
  const leftIcon = originalLeftIcon
    ? React.cloneElement(originalLeftIcon, { color: textColor[type] })
    : originalLeftIcon;
  const rightIcon = originalRightIcon
    ? React.cloneElement(originalRightIcon, { color: textColor[type] })
    : originalRightIcon;
  let justifyContent = layout.default;
  if (leftIcon != null) {
    justifyContent = layout.flexStart;
  }
  if (rightIcon != null || sublabel != null) {
    justifyContent = layout.spaceBetween;
  }

  let testID;
  if (testIDProps != null) {
    testID = `ButtonInner${testIDProps}`;
  }

  return (
    <View
      style={[
        styleSheet.buttonWrapper,
        disabled && styleSheet.disabled,
        theme(type).wrapper,
        justifyContent,
        circled && styleSheet.buttonCircled,
      ]}
      testID={testID}
    >
      <View style={layout.row}>
        {leftIcon != null && <View style={layout.rightSpace}>{leftIcon}</View>}
        {children ||
          (label != null && <ButtonTitle text={label} type={type} />)}
      </View>
      <View style={layout.row}>
        {sublabel != null && (
          <View style={layout.row}>
            <ButtonTitle text={sublabel} type={type} variant="sublabel" />
          </View>
        )}
        {rightIcon != null && <View style={layout.leftSpace}>{rightIcon}</View>}
      </View>
    </View>
  );
}

const layout = StyleSheet.create({
  default: {
    justifyContent: 'center',
  },
  spaceBetween: {
    justifyContent: 'space-between',
    web: {
      justifyContent: 'center',
    },
  },
  flexStart: {
    justifyContent: 'flex-start',
    web: {
      justifyContent: 'center',
    },
  },
  row: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  leftSpace: {
    marginStart: 5,
  },
  rightSpace: {
    marginEnd: 14,
  },
});

const styleSheet = StyleSheet.create({
  buttonWrapper: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    paddingHorizontal: parseInt(defaultTokens.spaceXSmall, 10),
    paddingVertical: 11,
    height: parseInt(defaultTokens.heightButtonNormal, 10),
    borderRadius: parseInt(defaultTokens.borderRadiusLarge, 10),
  },
  buttonCircled: {
    borderRadius: parseInt(defaultTokens.heightButtonNormal, 10) / 2,
  },
  disabled: {
    opacity: 0.5,
  },
});

const theme = (type: ButtonType = 'primary') =>
  StyleSheet.create({
    wrapper: {
      backgroundColor: wrapperColor[type],
    },
  });
