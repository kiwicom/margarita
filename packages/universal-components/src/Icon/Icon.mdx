---
name: Icon
menu: Components
route: /icon
---

import { Playground, PropsTable } from 'docz'
import { View } from 'react-native'
import { Icon } from '.'
import IconsList from './IconsList'


# Icon

TODO

## Properties

<PropsTable of={Icon} />

## Basic usage

<Playground>
  <Icon name="check" />
</Playground>

## Icons List

<IconsList />

