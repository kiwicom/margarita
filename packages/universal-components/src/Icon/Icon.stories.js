// @flow

import * as React from 'react';
import { storiesOf } from '@storybook/react-native';

import IconsList from './IconsList';

import { Icon } from '.';

storiesOf('Icon', module)
  .addDecorator(getStory => getStory())
  .add('Regular Icon', () => (
    <React.Fragment>
      <Icon name="check" />
    </React.Fragment>
  ))
  .add('Custom Icon', () => (
    <React.Fragment>
      <Icon name="check" color="#46B655" size="large" />
    </React.Fragment>
  ))
  .add('Wrong Icon Name', () => (
    <React.Fragment>
      {/* $FlowExpectedError: Wrong Icon name for showcase purposes */}
      <Icon name="-" color="#46B655" size="large" />
    </React.Fragment>
  ))
  .add('All Icons list', () => (
    <React.Fragment>
      <IconsList />
    </React.Fragment>
  ));
