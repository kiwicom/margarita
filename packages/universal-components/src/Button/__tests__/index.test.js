// @flow

import * as React from 'react';
import { Platform, View } from 'react-native';
import { render, shallow, fireEvent } from 'react-native-testing-library';
import snapshotDiff from 'snapshot-diff';

import ButtonWeb from '../Button.web';
import ButtonNative from '../Button.native';
import ButtonInner from '../ButtonInner';
import { Icon } from '../../Icon';

const originalPlatform = Platform.OS;
const noop = jest.fn();
const buttonText = 'Button text';
const buttonSublabel = 'Button sublabel';
const leftIcon = <Icon name="calendar" />;
const rightIcon = <Icon name="chevron-right" />;
const width = 200;
const disabled = false;
const href = 'https://lolilol.com';
const block = false;
const type = 'info';
const testID = 'testID';
const circled = false;

afterEach(() => {
  Platform.OS = originalPlatform;
});

describe('Button - web', () => {
  Platform.OS = 'web';
  const { getByText, getAllByProps } = render(
    <ButtonWeb
      onPress={noop}
      type={type}
      sublabel={buttonSublabel}
      leftIcon={leftIcon}
      rightIcon={rightIcon}
      width={width}
      disabled={disabled}
      href={href}
      block={block}
      testID={testID}
      label={buttonText}
      circled={circled}
    />,
  );

  it('should have the right text', () => {
    expect(getByText(buttonText)).toBeDefined();
  });

  it('should not have any sublabel', () => {
    expect(() => getByText(buttonSublabel)).toThrow('No instances found');
  });

  it('should have passed props', () => {
    expect(
      getAllByProps({
        onPress: noop,
        type,
        sublabel: buttonSublabel,
        leftIcon,
        rightIcon,
        width,
        disabled,
        href,
        block,
        testID,
        circled,
      }),
    ).toBeDefined();
  });

  it('should render a button by default', () => {
    const defaultButton = render(
      <ButtonWeb onPress={noop}>{buttonText}</ButtonWeb>,
    );
    // $FlowFixMe react-native-testing-library does not allow HTML elements
    expect(defaultButton.getByType('button')).toBeDefined();
    // $FlowFixMe react-native-testing-library does not allow HTML elements
    fireEvent(defaultButton.getByType('button'), 'click');
  });

  it('should render an <a> tag if href is provided', () => {
    const withHref = render(
      <ButtonWeb onPress={noop} href={href}>
        {buttonText}
      </ButtonWeb>,
    );
    // $FlowFixMe react-native-testing-library does not allow HTML elements
    expect(withHref.getByType('a')).toBeDefined();
    // $FlowFixMe react-native-testing-library does not allow HTML elements
    expect(() => fireEvent(withHref.getByType('a'), 'click')).toThrow(
      'No handler function found for event: click',
    );
  });

  it('should match the snapshot', () => {
    const { output } = shallow(
      <ButtonWeb
        onPress={noop}
        type={type}
        sublabel={buttonSublabel}
        leftIcon={leftIcon}
        rightIcon={rightIcon}
        width={width}
        disabled={disabled}
        href={href}
        block={block}
        circled={circled}
      >
        {buttonText}
      </ButtonWeb>,
    );
    expect(output).toMatchSnapshot();
  });

  it('should match snapshot diff', () => {
    const extend = render(
      <ButtonWeb
        onPress={noop}
        type={type}
        sublabel={buttonSublabel}
        leftIcon={leftIcon}
        rightIcon={rightIcon}
        width={width}
        disabled={disabled}
        href={href}
        block={block}
        circled={circled}
      >
        {buttonText}
      </ButtonWeb>,
    );
    const base = render(<ButtonWeb onPress={noop}>{buttonText}</ButtonWeb>);

    expect(snapshotDiff(base, extend)).toMatchSnapshot();
  });
});

describe('Button - native', () => {
  Platform.OS = 'ios';
  const { getByText, getByType, getAllByProps } = render(
    <ButtonNative
      onPress={noop}
      type={type}
      sublabel={buttonSublabel}
      leftIcon={leftIcon}
      rightIcon={rightIcon}
      width={width}
      disabled={disabled}
      href={href}
      block={block}
      testID={testID}
      label={buttonText}
      circled={circled}
    />,
  );

  it('should have the right text', () => {
    expect(getByText(buttonText)).toBeDefined();
  });

  it('should have the right sublabel', () => {
    expect(getByText(buttonSublabel)).toBeDefined();
  });

  it('should execute onPress method', () => {
    fireEvent(getByType(ButtonNative), 'press');
  });

  it('should have passed props', () => {
    expect(
      getAllByProps({
        onPress: noop,
        type,
        sublabel: buttonSublabel,
        leftIcon,
        rightIcon,
        width,
        disabled,
        href,
        block,
        testID,
        circled,
      }),
    ).toBeDefined();
  });

  it('should match the snapshot', () => {
    const { output } = shallow(
      <ButtonNative
        onPress={noop}
        type={type}
        sublabel={buttonSublabel}
        leftIcon={leftIcon}
        rightIcon={rightIcon}
        width={width}
        disabled={disabled}
        href={href}
        block={block}
        circled={circled}
      >
        {buttonText}
      </ButtonNative>,
    );
    expect(output).toMatchSnapshot();
  });

  it('should match snapshot diff', () => {
    const extend = render(
      <ButtonNative
        onPress={noop}
        type={type}
        sublabel={buttonSublabel}
        leftIcon={leftIcon}
        rightIcon={rightIcon}
        width={width}
        disabled={disabled}
        href={href}
        block={block}
        circled={circled}
      >
        {buttonText}
      </ButtonNative>,
    );
    const base = render(
      <ButtonNative onPress={noop}>{buttonText}</ButtonNative>,
    );
    expect(snapshotDiff(base, extend)).toMatchSnapshot();
  });
});

describe('ButtonInner', () => {
  it('should render children instead of label if passed', () => {
    const label = 'should not render';
    const wrapper = render(
      <ButtonInner label={label}>
        <View testID="child-view" />
      </ButtonInner>,
    );
    expect(wrapper.getByTestId('child-view')).toBeDefined();
    expect(() => {
      wrapper.getByText(label);
    }).toThrow();
  });

  it('should render label if no children are passed', () => {
    const wrapper = render(<ButtonInner label="test" />);
    expect(wrapper.getByText('test')).toBeDefined();
  });
});
