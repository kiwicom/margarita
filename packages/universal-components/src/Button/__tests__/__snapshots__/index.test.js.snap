// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button - native should match snapshot diff 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -8,11 +8,11 @@
    onResponderTerminationRequest={[Function bound touchableHandleResponderTerminationRequest]}
    onStartShouldSetResponder={[Function bound touchableHandleStartShouldSetResponder]}
    style={
      Object {
        \\"opacity\\": 1,
-       \\"width\\": undefined,
+       \\"width\\": 200,
      }
    }
  >
    <View>
      <View
@@ -27,35 +27,127 @@
              \\"paddingHorizontal\\": 8,
              \\"paddingVertical\\": 11,
            },
            false,
            Object {
-             \\"backgroundColor\\": \\"#00a991\\",
+             \\"backgroundColor\\": \\"#e0f6ff\\",
            },
            Object {
-             \\"justifyContent\\": \\"center\\",
+             \\"justifyContent\\": \\"space-between\\",
            },
            false,
          ]
        }
      >
        <View
          style={
            Object {
              \\"alignItems\\": \\"center\\",
              \\"flexDirection\\": \\"row\\",
+           }
          }
+       >
+         <View
+           style={
+             Object {
+               \\"marginEnd\\": 14,
+             }
+           }
+         >
+           <Text
+             style={
+               Array [
+                 Object {
+                   \\"fontFamily\\": \\"orbit-icons\\",
+                 },
+                 Object {
+                   \\"fontSize\\": 24,
+                   \\"height\\": 24,
+                   \\"lineHeight\\": 24,
+                   \\"width\\": 24,
+                 },
+                 Object {
+                   \\"color\\": \\"#0176D2\\",
+                 },
+                 undefined,
+               ]
              }
            >
+             f
+           </Text>
+         </View>
          Button text
        </View>
        <View
          style={
            Object {
              \\"alignItems\\": \\"center\\",
              \\"flexDirection\\": \\"row\\",
            }
          }
-       />
+       >
+         <View
+           style={
+             Object {
+               \\"alignItems\\": \\"center\\",
+               \\"flexDirection\\": \\"row\\",
+             }
+           }
+         >
+           <Text
+             style={
+               Array [
+                 Object {
+                   \\"fontFamily\\": \\"Roboto\\",
+                   \\"margin\\": 0,
+                 },
+                 Array [
+                   Object {},
+                   Object {
+                     \\"fontSize\\": 14,
+                     \\"fontWeight\\": \\"normal\\",
+                     \\"lineHeight\\": 18,
+                   },
+                   Object {
+                     \\"color\\": \\"#0176D2\\",
+                   },
+                   undefined,
+                 ],
+               ]
+             }
+           >
+             Button sublabel
+           </Text>
+         </View>
+         <View
+           style={
+             Object {
+               \\"marginStart\\": 5,
+             }
+           }
+         >
+           <Text
+             style={
+               Array [
+                 Object {
+                   \\"fontFamily\\": \\"orbit-icons\\",
+                 },
+                 Object {
+                   \\"fontSize\\": 24,
+                   \\"height\\": 24,
+                   \\"lineHeight\\": 24,
+                   \\"width\\": 24,
+                 },
+                 Object {
+                   \\"color\\": \\"#0176D2\\",
+                 },
+                 undefined,
+               ]
+             }
+           >
+             
+           </Text>
+         </View>
+       </View>
      </View>
    </View>
  </View>"
`;

exports[`Button - native should match the snapshot 1`] = `
<Button
  disabled={false}
  onPress={
    [MockFunction] {
      "calls": Array [
        Array [
          undefined,
        ],
        Array [
          undefined,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  width={200}
>
  <ButtonInner
    circled={false}
    disabled={false}
    leftIcon={
      <Icon
        name="calendar"
      />
    }
    rightIcon={
      <Icon
        name="chevron-right"
      />
    }
    sublabel="Button sublabel"
    type="info"
  >
    Button text
  </ButtonInner>
</Button>
`;

exports[`Button - web should match snapshot diff 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,21 +1,19 @@
- <button
-   disabled={false}
-   onClick={[Function mockConstructor]}
+ <a
+   href=\\"https://lolilol.com\\"
    style={
      Object {
        \\"backgroundColor\\": \\"transparent\\",
        \\"border\\": 0,
        \\"cursor\\": \\"pointer\\",
        \\"display\\": \\"inline-block\\",
        \\"padding\\": 0,
        \\"pointerEvents\\": \\"auto\\",
        \\"textDecoration\\": \\"none\\",
-       \\"width\\": \\"inherit\\",
+       \\"width\\": 200,
      }
    }
-   type=\\"button\\"
  >
    <View
      style={
        Array [
          Object {
@@ -27,34 +25,93 @@
            \\"paddingHorizontal\\": 8,
            \\"paddingVertical\\": 11,
          },
          false,
          Object {
-           \\"backgroundColor\\": \\"#00a991\\",
+           \\"backgroundColor\\": \\"#e0f6ff\\",
          },
          Object {
-           \\"justifyContent\\": \\"center\\",
+           \\"justifyContent\\": \\"space-between\\",
          },
-         undefined,
+         false,
        ]
      }
    >
      <View
        style={
          Object {
            \\"alignItems\\": \\"center\\",
            \\"flexDirection\\": \\"row\\",
+         }
+       }
+     >
+       <View
+         style={
+           Object {
+             \\"marginEnd\\": 14,
            }
          }
        >
+         <Text
+           style={
+             Array [
+               Object {
+                 \\"fontFamily\\": \\"orbit-icons\\",
+               },
+               Object {
+                 \\"fontSize\\": 24,
+                 \\"height\\": 24,
+                 \\"lineHeight\\": 24,
+                 \\"width\\": 24,
+               },
+               Object {
+                 \\"color\\": \\"#0176D2\\",
+               },
+               undefined,
+             ]
+           }
+         >
+           f
+         </Text>
+       </View>
        Button text
      </View>
      <View
        style={
          Object {
            \\"alignItems\\": \\"center\\",
            \\"flexDirection\\": \\"row\\",
          }
        }
-     />
+     >
+       <View
+         style={
+           Object {
+             \\"marginStart\\": 5,
+           }
+         }
+       >
+         <Text
+           style={
+             Array [
+               Object {
+                 \\"fontFamily\\": \\"orbit-icons\\",
+               },
+               Object {
+                 \\"fontSize\\": 24,
+                 \\"height\\": 24,
+                 \\"lineHeight\\": 24,
+                 \\"width\\": 24,
+               },
+               Object {
+                 \\"color\\": \\"#0176D2\\",
+               },
+               undefined,
+             ]
+           }
+         >
+           
+         </Text>
+       </View>
      </View>
- </button>
+   </View>
+ </a>"
`;

exports[`Button - web should match the snapshot 1`] = `
<a
  href="https://lolilol.com"
  style={
    Object {
      "backgroundColor": "transparent",
      "border": 0,
      "cursor": "pointer",
      "display": "inline-block",
      "padding": 0,
      "pointerEvents": "auto",
      "textDecoration": "none",
      "width": 200,
    }
  }
>
  <ButtonInner
    circled={false}
    disabled={false}
    leftIcon={
      <Icon
        name="calendar"
      />
    }
    rightIcon={
      <Icon
        name="chevron-right"
      />
    }
    type="info"
  >
    Button text
  </ButtonInner>
</a>
`;
